### PR TITLE
tests/DCPS/UnitTests does not conform to guidelines

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

* Unit tests under /tests/DCPS/UnitTests do not
  /docs/internal/unit_tests.rst.

* /tests/DCPS/UnitTests contains an integration test.

* /tests/DCPS/UnitTests contains a stress test.

Solution
--------

* Move unit tests to /tests/unit-tests.

* Move stress test to /tests/stress-tests.

* Move integratin tests to /tests/DCPS/BitDataReader.

* Provide a generic run_test.pl for launching unit and stress tests.

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -4820,14 +4820,14 @@ jobs:
...
+        cd tests\unit-tests
         perl %ACE_ROOT%\bin\mwc.pl -type vs2019 -static -features no_rapidjson=1 -features ipv6=1 -features ssl=1 -features no_cxx11=0
     - name: make unit tests
       shell: cmd
       run: |
         cd OpenDDS
         call setenv.cmd
-        cd tests\DCPS\UnitTests
+        cd tests\unit-tests
         msbuild -p:Configuration=Debug,Platform=x64 -m UnitTests.sln
```
Comment (jrw972) [inline .github/workflows/build_and_test.yml:4831]: @ClaytonCalabrese @mitza-oci There are multiple mpc files in `tests/unit-tests`.  How should I update the line that contains `UnitTests.sln`?

Comment (simpsont-oci) [reply .github/workflows/build_and_test.yml:4831]: If there are multiple .sln files from the previous step, I suppose we just iterate over them? I'm not on a windows machine to test, but maybe something like:
`for %f in (.\*.sln) do msbuild -p:Configuration=Debug,Platform=x64 -m %f`

It's possible that passing *.sln to msbuild might work as well...

Comment (mitza-oci) [reply .github/workflows/build_and_test.yml:4831]: Why would there be multiple workspaces?

File: tests/auto_run_tests.pl [hunk 2]
```diff
@@ -73,9 +73,13 @@
...
         $test =~ s!/./!/!g;
 
-        if ($test =~ /(.*)\/([^\/]*)$/) {
+        my @tokens = split / /, $test;
+        my $progname = shift @tokens;
+
+        if ($progname =~ /(.*)\/([^\/]*)$/) {
             $directory = $1;
-            $program = $2;
+            push(@tokens, ($2));
```
Comment (mitza-oci) [inline tests/auto_run_tests.pl:81]:
```suggestion
            push(@tokens, $2);
```

File: tests/run_test.pl [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+if (! -e $executable) {
+    print STDERR "Executable $executable does not exist\n";
+    # FUTURE: Return a status indicating that the test was not run.
+    exit 0;
+}
+
+my $test = new PerlDDS::TestFramework();
+$test->process("$executable", "$executable", "");
```
Comment (mitza-oci) [inline tests/run_test.pl:32]:
```suggestion
$test->process($executable, $executable, "");
```

File: tests/run_test.pl [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+print STDERR "Running $executable\n";
+$test->start_process("$executable");
```
Comment (mitza-oci) [inline tests/run_test.pl:34]:
```suggestion
$test->start_process($executable);
```

File: tests/unit-tests/dds/DCPS/RTPS/ParameterListConverter.cpp [hunk 5]
```diff
@@ -496,13 +665,37 @@ Parameter get(const ParameterList& param_list,
...
+  ASSERT_EQ(0, strcmp("Property 0", w2.ddsParticipantDataSecure.identity_status_token.properties[0].name));
+  ASSERT_EQ(0, strcmp("PropertyValue 0", w2.ddsParticipantDataSecure.identity_status_token.properties[0].value));
+  ASSERT_EQ(0, strcmp("BinaryProperty 0", w2.ddsParticipantDataSecure.identity_status_token.binary_properties[0].name));
+  ASSERT_EQ(0, memcmp("BinaryPropertyValue 0", w2.ddsParticipantDataSecure.identity_status_token.binary_properties[0].value.get_buffer(), strlen("BinaryPropertyValue 0")));
+}
+#endif
+
 int
-ACE_TMAIN(int, ACE_TCHAR*[])
+ACE_TMAIN(int argc, ACE_TCHAR* argv[])
```
Comment (mitza-oci) [inline tests/unit-tests/dds/DCPS/RTPS/ParameterListConverter.cpp:693]: Use `ACE_UNUSED_ARG`

File: tests/unit-tests/dds/DCPS/RTPS/UnitTests.mpc [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+  Source_Files {
+    BaseMessageUtils.cpp
+  }
+}
+
+project(*ParameterListConverter): dcps_rtpsexe, dcps_test, googletest {
+  exename   = *
+  //  requires += no_opendds_safety_profile
+  //  avoids += no_opendds_security
```
Comment (mitza-oci) [inline tests/unit-tests/dds/DCPS/RTPS/UnitTests.mpc:12]: ?

File: tools/scripts/stress_tests.pl [hunk 7]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env perl
```
Comment (mitza-oci) [inline tools/scripts/stress_tests.pl:1]: Looks like this script could be refactored to reduce duplication with the other one

**Review by mitza-oci — CHANGES_REQUESTED** [review]
New scripts are failing

**Review by mitza-oci — APPROVED** [review]

File: tests/run_test.pl [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+if ($executable eq '') {
+    print STDERR "ERROR: no executable was specified\n";
+    exit 1;
+}
+
+if (! -e $executable) {
+    print STDERR "Executable $executable does not exist\n";
+    # FUTURE: Return a status indicating that the test was not run.
+    exit 0;
+}
```
[the remark this answers is not in the document]
Comment (mitza-oci) [reply tests/run_test.pl:29]: We still need to figure this one out.  It's not going to work to make an assumption that a file with a name of the test exists in this directory.  Putting things in different directories (`-ExeSubDir`) is possible on all platforms (like `ARCH` with gnuace and some of the alternate configurations on MSVC).  Also there are suffixes like `.exe` to take into account.  The common Perl modules (both in OpenDDS and in ACE) have some functions that are useful for dealing with this.  But the larger question is what we want it to do.

**Review by mitza-oci — CHANGES_REQUESTED** [review]
see notes on existing comment

**Review by mitza-oci — APPROVED** [review]

File: MPC/config/opendds_cxx11.mpb [hunk 9]
```diff
@@ -2,6 +2,7 @@ project: dcps, dcps_ts_defaults {
 
   avoids += no_cxx11
   requires += no_opendds_safety_profile
+  macros += OPENDDS_HAX_CXX11
```
Comment (mitza-oci) [inline MPC/config/opendds_cxx11.mpb:5]:
```suggestion
  macros += OPENDDS_HAS_CXX11
```

Comment (mitza-oci) [reply MPC/config/opendds_cxx11.mpb:5]: and fix other occurrences of this macro

File: tests/dcps_tests.lst [hunk 10]
```diff
@@ -237,9 +237,12 @@ tests/DCPS/EntityLifecycleStress/run_test.pl rtps_disc publishers 10 subscribers
 tests/DCPS/LivelinessKeepAliveTest/run_test.pl: !DCPS_MIN !OPENDDS_SAFETY_PROFILE
 tests/DCPS/LivelinessTimeout/run_test.pl: !DCPS_MIN !GH_ACTIONS
 tests/DCPS/LivelinessTimeout/run_test.pl rtps_disc: !DCPS_MIN RTPS
-tests/DCPS/UnitTests/run_test.pl: !DCPS_MIN !GH_ACTIONS
-tests/unit-tests/dds/DCPS/run_test_unit.pl: !DCPS_MIN !GH_ACTIONS
-tests/unit-tests/dds/DCPS/transport/framework/run_test_unit.pl
+tests/DCPS/BitDataReader/run_test.pl: !DCPS_MIN !GH_ACTIONS
```
Comment (mitza-oci) [inline tests/dcps_tests.lst:240]: Does it need to have `!GH_ACTIONS`?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: tests/dcps_tests.lst (hunk 10 above)
Comment (jrw972) [reply tests/dcps_tests.lst:240]: I went with what was there.  I can remove it and see what happens.

**Review by mitza-oci — APPROVED** [review]

File: .github/workflows/build_and_test.yml [hunk 11]
```diff
@@ -5311,6 +5311,17 @@ jobs:
...
+    - name: setup gtest
+      shell: cmd
+      run: |
+        cd OpenDDS/tests/googletest
+        git submodule init
+        git submodule update
+        mkdir build
+        cd build
+        mkdir install
+        cmake -DCMAKE_INSTALL_PREFIX=./install -DBUILD_SHARED_LIBS=ON ..
```
Comment (simpsont-oci) [inline .github/workflows/build_and_test.yml:5323]: I think these flags (and the flags for the other broken-apart "test" jobs) need to match (one way or the other) the flags used in the build step above which currently don't use BUILD_SHARED_LIBS=ON

**Pull request merged**